### PR TITLE
Update changelog

### PR DESCRIPTION
**Pull request by ralphtheninja**
Closes https://github.com/Level/abstract-leveldown/issues/179

~~WIP Don't merge~~

Go through all versions below

- [x] `0.0.0`
- [x] `0.0.1`
- [x] `0.0.2`
- [x] `0.1.0`
- [x] `0.2.0`
- [x] `0.2.1`
- [x] `0.2.2`
- [x] `0.2.3`
- [x] `0.3.0`
- [x] `0.4.0`
- [x] `0.4.0-1`
- [x] `0.5.0`
- [x] `0.6.0`
- [x] `0.6.1`
- [x] `0.7.0`
- [x] `0.7.1`
- [x] `0.7.2`
- [x] `0.7.3`
- [x] `0.7.4`
- [x] `0.8.0`
- [x] `0.8.1`
- [x] `0.8.2`
- [x] `0.9.0`
- [x] `0.10.0`
- [x] `0.10.1`
- [x] `0.10.2`
- [x] `0.11.0`
- [x] `0.11.1`
- [x] `0.11.2`
- [x] `0.11.3`
- [x] `0.11.4`
- [x] `0.12.0`
- [x] `0.12.1`
- [x] `0.12.2`
- [x] `0.12.3`
- [x] `0.12.4`
- [x] `1.0.0`
- [x] `2.0.0`
- [x] `2.0.1`
- [x] `2.0.2`
- [x] `2.0.3`
- [x] `2.1.0`
- [x] `2.1.1`
- [x] `2.1.2`
- [x] `2.1.3`
- [x] `2.1.4`
- [x] `2.2.0`
- [x] `2.2.1`
- [x] `2.2.2`
- [x] `2.3.0`
- [x] `2.3.1`
- [x] `2.4.0`
- [x] `2.4.1`
- [x] `2.5.0 `
- [x] `2.6.0`
- [x] `2.6.1`
- [x] `2.6.2`
- [x] `2.6.3`
- [x] `2.7.0`
- [x] `2.7.1`
- [x] `2.7.2`
- [x] `3.0.0`
- [ ] ~~add upgrade guide to coming v4 release, i.e. `Unreleased` section~~
- [x] use `Added`, `Changed`, `Removed` and `Fixed` appropriately for each version, see https://github.com/olivierlacan/keep-a-changelog/blob/master/CHANGELOG.md

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -19,109 +23,136 @@
...
+[2.2.0]: https://github.com/level/abstract-leveldown/compare/v2.1.4...v2.2.0
+[2.1.4]: https://github.com/level/abstract-leveldown/compare/v2.1.3...v2.1.4
+[2.1.3]: https://github.com/level/abstract-leveldown/compare/v2.1.2...v2.1.3
+[2.1.2]: https://github.com/level/abstract-leveldown/compare/v2.1.1...v2.1.2
+[2.1.1]: https://github.com/level/abstract-leveldown/compare/v2.1.0...v2.1.1
+[2.1.0]: https://github.com/level/abstract-leveldown/compare/v2.0.3...v2.1.0
+[2.0.3]: https://github.com/level/abstract-leveldown/compare/v2.0.2...v2.0.3
+[2.0.2]: https://github.com/level/abstract-leveldown/compare/v2.0.1...v2.0.2
+[2.0.1]: https://github.com/level/abstract-leveldown/compare/v2.0.0...v2.0.1
+[2.0.0]: https://github.com/level/abstract-leveldown/compare/0.11.1...v2.0.0
```
Comment (ralphtheninja) [inline CHANGELOG.md]: I'm aware of the big gap here. Will adjust later once I have added the missing versions/tags.

**Comment by ralphtheninja** [comment]
@vweevers I just noticed something odd in the git history of `abstract-leveldown`, see screenshot below

weird-git-history

There's a remote `patch-1` branch which contains tags `v1.1.0` to `v1.8.1` which does _not_ correspond to any existing versions of `abstract-leveldown`. We only released `v1.0.0` before jumping directly to `v2.0.0`.

**Comment by ralphtheninja** [comment]
Seems to be releated to `AbstractNoSQL`. I'm wondering if @snowyu accidentally pushed to the wrong remote or something like that.

**Comment by vweevers** [comment]
😮 If we don't get a timely response from @snowyu, let's delete that branch.

File: CHANGELOG.md [hunk 2]
```diff
@@ -151,8 +287,32 @@ No change.
...
+[2.0.0]: https://github.com/level/abstract-leveldown/compare/v1.0.0...v2.0.0
+[1.0.0]: https://github.com/level/abstract-leveldown/compare/v0.12.4...v1.0.0
+[0.12.4]: https://github.com/level/abstract-leveldown/compare/v0.12.3...v0.12.4
+[0.12.3]: https://github.com/level/abstract-leveldown/compare/v0.12.2...v0.12.3
+[0.12.2]: https://github.com/level/abstract-leveldown/compare/v0.12.1...v0.12.2
+[0.12.1]: https://github.com/level/abstract-leveldown/compare/v0.12.0...v0.12.1
+[0.12.0]: https://github.com/level/abstract-leveldown/compare/v0.11.4...v0.12.0
+[0.11.4]: https://github.com/level/abstract-leveldown/compare/v0.11.3...v0.11.4
+[0.11.3]: https://github.com/level/abstract-leveldown/compare/v0.11.2...v0.11.3
+[0.11.2]: https://github.com/level/abstract-leveldown/compare/0.11.1...v0.11.2
```
Comment (ralphtheninja) [inline CHANGELOG.md]: Early tags don't start with a `v` which is a bit annoying.

**Comment by ralphtheninja** [comment]
> If we don't get a timely response from @snowyu, let's delete that branch.

@vweevers I pushed up a branch named `THIS_BRANCH_WILL_BE_DELETED` and removed all the tags. If you have any local tags, make sure you don't do `git push --tags` or similar.

**Comment by ralphtheninja** [comment]
@vweevers Check current state at https://github.com/Level/abstract-leveldown/blob/changelog/CHANGELOG.md

**Comment by ralphtheninja** [comment]
A suggestion is to use links for all contributors, e.g. @vweevers would be wrapped with a link to your github page.

**Comment by ralphtheninja** [comment]
@vweevers Ok, done with `Unreleased` as well. *wipes sweat from his forehead* :grinning: I think this changelog is starting to look pretty bad ass actually.

**Comment by vweevers** [comment]
Perfect. This was done with care, with great attention to detail and consistency. 

I fully agree a changelog in this form is valuable to have, separate from a upgrade guide yet more human than commit logs.

This serves as a blueprint for all changelogs. I love the little Historical Notes. Acknowledging mistakes goes a long way.

Couple of questions:

- What tools did you use for this?
- Should we make an effort to write better commit messages, so we can automate the changelog (more)?
- Which commits did you exclude? Any recommendations we can follow in the future?

**Comment by vweevers** [comment]
> Should we make an effort to write better commit messages

Better meaning more consistent and to the point.

**Comment by ralphtheninja** [comment]
> What tools did you use for this?

Pure manual labor and emacs :)

> Should we make an effort to write better commit messages, so we can automate the changelog (more)?

I thought about this. There's a bunch of trade offs to be made here. If we go down the path of forcing a certain type of commit messages, then changelog automation is trivial ([generate-changelog](https://github.com/lob/generate-changelog) does this), but once in a while a commit message will slip through that doesn't follow the rules and then automation will be problematic. Also, I'm not sure it's beneficial to enforce a certain commit style, for the sole purpose of automation. That said, we could always try to write better commit messages (i.e. not "changed some stuff") even though they aren't following a predetermined format.

Maybe it's better to have a simple tool that just checks the current changelog and makes sure it has an entry for respective release. E.g. lets say that I create a new release (new tag and new npm release) but forget to write an entry in the changelog, then it would be trivial to have tool check for this entry, and have tests fail or something like that.

Once a changelog exists, it's not a lot of work to either update it at each release, or update it iteratively as new commits come in.

> Which commits did you exclude? Any recommendations we can follow in the future?

I went through the whole history and all the diffs and while doing this I've noticed commits that undoes previous commits. For example, someone commits a change, then gets feedback and makes more changes that removes or modifies the previous commit. I decided to not document this, but rather document the final state of the contribution.

**Comment by ralphtheninja** [comment]
@vweevers I had an idea the other day about a `changelogify` module, which could help out with either generating a changelog template, i.e. grab versions from npm and get tags from git, and use that information to generate a starting changelog that has all the fields filled in. It could generate all the `compare`-links and whatnot.

Another idea could be to create `verify-changelog` that could be used as a testing tool.

**Pull request merged**